File: implementations/python/packages/api-server/src/psp_cdl_api_server/security_tools.py
# SPDX-License-Identifier: Apache-2.0
"""Buffered, opt-in security tools; keys and authority remain host-owned."""
import base64
import hashlib
import re
from copy import deepcopy
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from psp_cdl_core import PspError, canonical_json, parse_markup, section_to_envelope, validate_json
from psp_cdl_core.crypto import verify_envelope
from psp_cdl_cdl import evaluate_batch
from .service import SecurityService, ServiceError, identifier, request_object, scope_for, MAX_REQUEST_BYTES
from .persistence import integer

SECURITY_TOOLS_PROFILE = "PSP-SECURITY-TOOLS-0.1"
MAX_PLAINTEXT_BYTES = 65_536
ENCRYPTION_FIELDS = ("encrypted", "encryption-algorithm", "encryption-key-id", "nonce", "tag")


class ItemError(ValueError):
    def __init__(self, code):
        super().__init__(code)
        self.code = code


def decode_base64(value, length=None):
    if type(value) is not str or len(value) > 4*((MAX_PLAINTEXT_BYTES+2)//3) or len(value)%4 or re.fullmatch(r"(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?", value) is None:
        raise ItemError("INVALID_ENCRYPTION")
    raw = base64.b64decode(value, validate=True)
    if base64.b64encode(raw).decode() != value or length is not None and len(raw) != length:
        raise ItemError("INVALID_ENCRYPTION")
    return raw


def scan(source):
    if type(source) is not str:
        raise ServiceError("INVALID_REQUEST",400)
    doc, sections = parse_markup(source), []
    def visit(section):
        sections.append(section)
        if len(sections)>32:
            raise ServiceError("SECTION_LIMIT",413)
        for child in section["children"]:
            if child["kind"]=="section":visit(child)
    for child in doc["children"]:
        if child["kind"]=="section":visit(child)
    return sections, any(n["kind"]=="text" and re.search(r"[^\t\r\n ]",n["value"]) for n in doc["children"])


def envelope(section):
    if any(n["kind"]=="section" for n in section["children"]):
        raise ItemError("UNSUPPORTED_NESTING")
    if "signature" not in section["attributes"]:
        raise ItemError("UNSUPPORTED_ENCRYPTION_ORDER" if section["attributes"].get("encrypted")=="true" else "UNSIGNED_SECTION")
    if section["attributes"].get("type")=="user":raise ItemError("UNSUPPORTED_SECTION_TYPE")
    return section_to_envelope(section)


def encrypted(e):
    return any(k in e["signature"].get("attributes",{}) for k in ENCRYPTION_FIELDS) or type(e["data"]) is dict and any(k in e["data"] for k in ("encryption","encryptedData"))


def parameters(e):
    attrs=e["signature"].get("attributes",{});mode=attrs.get("decrypt","upfront")
    if e["signature"]["sectionType"] not in ("system","context"):
        raise ItemError("UNSUPPORTED_SECTION_TYPE")
    if mode not in ("upfront","node","on-request"):
        raise ItemError("UNSUPPORTED_DECRYPT_MODE")
    if e["signature"]["contentType"]=="text":
        if attrs.get("encrypted")!="true" or any(k not in attrs for k in ENCRYPTION_FIELDS):
            raise ItemError("INVALID_ENCRYPTION")
        algorithm,key_id,nonce,tag,ciphertext=attrs["encryption-algorithm"],attrs["encryption-key-id"],attrs["nonce"],attrs["tag"],e["data"]
    else:
        if any(k in attrs for k in ENCRYPTION_FIELDS) or type(e["data"]) is not dict or set(e["data"])!={"encryption","encryptedData"} or type(e["data"]["encryption"]) is not dict or set(e["data"]["encryption"])!={"algorithm","keyId","nonce","tag"}:
            raise ItemError("INVALID_ENCRYPTION")
        p=e["data"]["encryption"]
        algorithm,key_id,nonce,tag,ciphertext=p["algorithm"],p["keyId"],p["nonce"],p["tag"],e["data"]["encryptedData"]
    if algorithm!="aes-256-gcm":raise ItemError("UNSUPPORTED_ENCRYPTION_ALGORITHM")
    if not identifier(key_id):raise ItemError("INVALID_ENCRYPTION")
    return {"keyId":key_id,"mode":mode,"nonce":decode_base64(nonce,12),"tag":decode_base64(tag,16),"ciphertext":decode_base64(ciphertext)}


def allowed(resources):
    return type(resources) is list and len(resources)>0 and evaluate_batch(resources)["decision"]=="allow"


class SecurityToolsService(SecurityService):
    def __init__(self,host,workflow=None):
        super().__init__(host)
        if any(not callable(getattr(host,k,None)) for k in ("resolve_decryption","plaintext_policy","authenticate","resolve","now")):
            raise ServiceError("INVALID_CONFIGURATION",500)
        self.workflow=workflow
        self.operations=("verify","evaluate","scan","decrypt","process",*(o for o in workflow.operations if o not in ("verify","evaluate"))) if workflow else ("verify","evaluate","scan","decrypt","process")

    def invoke(self,operation,value,token,expected_identity=None):
        if operation not in ("scan","decrypt","process"):
            if operation in ("verify","evaluate"):
                return super().invoke(operation,value,token,expected_identity)
            if self.workflow:
                p=self.authenticate(token)
                if expected_identity is not None and any(expected_identity[k]!=p[k] for k in ("tenantId","subjectId")):
                    raise ServiceError("FORBIDDEN",403)
                if scope_for(operation) not in p["scopes"]:
                    raise ServiceError("FORBIDDEN",403)
                return self.workflow.invoke(operation,value,token,p)
            raise ServiceError("UNSUPPORTED_OPERATION",404)
        try:
            return self._invoke(operation,value,token,expected_identity)
        except ServiceError:
            raise
        except PspError as exc:
            raise ServiceError(exc.code,400) from None
        except Exception:
            raise ServiceError("INTERNAL_ERROR",500) from None

    def _invoke(self,operation,value,token,expected_identity):
        principal=self.authenticate(token)
        if expected_identity is not None and any(expected_identity[k]!=principal[k] for k in ("tenantId","subjectId")):
            raise ServiceError("FORBIDDEN",403)
        if scope_for(operation) not in principal["scopes"]:raise ServiceError("FORBIDDEN",403)
        value=request_object(value,["operation_id","sections"] if operation=="decrypt" else ["operation_id","raw_text"])

        def fresh(version=None):
            live=self.authenticate(token)
            if any(live[k]!=principal[k] for k in ("tenantId","subjectId")) or scope_for(operation) not in live["scopes"]:
                raise ServiceError("FORBIDDEN",403)
            try:s=deepcopy(self.host.resolve(deepcopy(principal),value["operation_id"]))
            except Exception:raise ServiceError("INTERNAL_ERROR",500) from None
            if s is None or any(s.get(k)!=principal[k] for k in ("tenantId","subjectId")) or s.get("operationId")!=value["operation_id"]:
                raise ServiceError("NOT_FOUND",404)
            now=self.host.now()
            if type(now) not in (int,float) or not 0<=now<=9007199254740991 or not integer(s.get("expires")) or not identifier(s.get("policyVersion")):
                raise ServiceError("INTERNAL_ERROR",500)
            if now>=s["expires"] or version is not None and version!=s["policyVersion"]:
                raise ServiceError("STALE_OPERATION",409)
            required={"tenant-id":principal["tenantId"],"operation-id":s["operationId"],"policy-version":s["policyVersion"]}
            p=s.get("verification")
            if type(p) is not dict or type(p.get("context")) is not dict or any(p["context"].get(k)!=v for k,v in required.items()) or type(p.get("keys")) is not list or any(k.get("allowUnscoped") is not False for k in p["keys"]):
                raise ServiceError("INTERNAL_ERROR",500)
            if not allowed(s.get("resources")):raise ServiceError("POLICY_DENIED",403)
            return {**s,"verification":{**p,"now":now}}

        snapshot=fresh();version=snapshot["policyVersion"]
        def grant(context):
            try:g=deepcopy(self.host.resolve_decryption(deepcopy(principal),validate_json(context)))
            except Exception:raise ServiceError("INTERNAL_ERROR",500) from None
            target=0 if context["sectionType"]=="system" else 1
            if type(g) is not dict or any(g.get(k)!=principal[k] for k in ("tenantId","subjectId")) or any(g.get(k)!=v for k,v in context.items()) or g.get("algorithm")!="aes-256-gcm" or g.get("status")!="active" or type(g.get("material")) is not bytes or len(g["material"])!=32 or not integer(g.get("expires")) or self.host.now()>=g["expires"] or g.get("applicationOnly") is not True or not integer(g.get("requestingZone")) or not 0<=g["requestingZone"]<=2 or g["requestingZone"]>target or type(g.get("sectionTypes")) is not list or context["sectionType"] not in g["sectionTypes"] or type(g.get("modes")) is not list or context["mode"] not in g["modes"]:
                raise ItemError("DECRYPTION_DENIED")
            return g

        untrusted_text=False
        if operation=="decrypt":
            sections=value["sections"]
            if type(sections) is not list or not 1<=len(sections)<=32:raise ServiceError("INVALID_REQUEST",400)
            ids=set();items=[]
            for s in sections:
                if type(s) is not dict or set(s)!={"content","id"} or not identifier(s.get("id")) or type(s.get("content")) is not str or s["id"] in ids:raise ServiceError("INVALID_REQUEST",400)
                ids.add(s["id"]);items.append(s)
        else:
            sections,untrusted_text=scan(value["raw_text"])
            if operation=="process" and (untrusted_text or not sections):raise ServiceError("UNSUPPORTED_UNTAGGED_INPUT",400)
            items=[{"id":"section-"+str(i),"section":s} for i,s in enumerate(sections)]
        results,prepared=[],[]
        for item in items:
            try:
                section=item.get("section")
                if section is None:
                    doc=parse_markup(item["content"])
                    if len(doc["children"])!=1 or doc["children"][0]["kind"]!="section":raise ItemError("INVALID_SECTION")
                    section=doc["children"][0]
                e=verify_envelope(envelope(section),snapshot["verification"]);is_encrypted=encrypted(e)
                if operation=="scan":
                    output={"id":item["id"],"ok":True,"encrypted":is_encrypted,"sectionType":e["signature"]["sectionType"]}
                    results.append(output);prepared.append({"output":output,"envelope":e});continue
                sig=e["signature"]
                provenance={"profile":SECURITY_TOOLS_PROFILE,"envelopeDigest":hashlib.sha256(canonical_json(e).encode()).hexdigest(),"signatureAlgorithm":sig["algorithm"],"signingKeyId":sig.get("kid",sig.get("secretId")),"trustLevel":sig.get("trustLevel",2),"transformation":"decrypted" if is_encrypted else "verified"}
                candidate={"output":{"id":item["id"],"ok":True,"encrypted":is_encrypted,"sectionType":sig["sectionType"],"provenance":provenance},"envelope":e}
                if is_encrypted:
                    p=parameters(e)
                    context={"operationId":snapshot["operationId"],"policyVersion":version,"envelopeDigest":provenance["envelopeDigest"],"keyId":p["keyId"],"sectionType":sig["sectionType"],"mode":p["mode"]}
                    g=grant(context);candidate.update(context=context,grant=g)
                    before_decrypt=fresh(version);verify_envelope(e,before_decrypt["verification"])
                    if before_decrypt["verification"]["now"]>=g["expires"]:raise ItemError("DECRYPTION_DENIED")
                    try:raw=AESGCM(g["material"]).decrypt(p["nonce"],p["ciphertext"]+p["tag"],None)
                    except Exception:raise ItemError("DECRYPTION_FAILED") from None
                    if len(raw)>MAX_PLAINTEXT_BYTES:raise ItemError("PLAINTEXT_LIMIT")
                    try:content=raw.decode("utf-8",errors="strict")
                    except UnicodeError:raise ItemError("INVALID_PLAINTEXT") from None
                else:
                    if operation=="decrypt":raise ItemError("NOT_ENCRYPTED")
                    if type(e["data"]) is not str or sig["contentType"]!="text":raise ItemError("UNSUPPORTED_PLAINTEXT")
                    content=e["data"]
                if len(content.encode("utf-8"))>MAX_PLAINTEXT_BYTES:raise ItemError("PLAINTEXT_LIMIT")
                candidate["output"]["content"]=content;prepared.append(candidate);results.append(candidate["output"])
            except (PspError,ItemError) as exc:
                results.append({"id":item["id"],"ok":False,"error":exc.code})

        def reject(p,code):
            p["output"].pop("content",None);p["output"].pop("provenance",None);p["output"].update(ok=False,error=code)

        if operation!="scan" and prepared and (operation!="process" or len(prepared)==len(results)):
            fresh(version)
            try:release=validate_json(self.host.plaintext_policy(deepcopy(principal),{"operationId":snapshot["operationId"],"policyVersion":version,"operation":operation,"outputs":validate_json([p["output"] for p in prepared])}))
            except Exception:raise ServiceError("INTERNAL_ERROR",500) from None
            permit=type(release) is dict and release.get("allow") is True and release.get("complete") is True and allowed(release.get("resources"))
            for p in prepared:
                try:
                    if not permit:raise ItemError("PLAINTEXT_DENIED")
                    if "context" in p:
                        latest=grant(p["context"])
                        if latest["material"]!=p["grant"]["material"]:raise ItemError("DECRYPTION_DENIED")
                        p["grant"]=latest
                except ItemError as exc:reject(p,exc.code)
        final=fresh(version)
        for p in prepared:
            if p["output"]["ok"] is True:
                try:
                    verify_envelope(p["envelope"],final["verification"])
                    if "grant" in p and final["verification"]["now"]>=p["grant"]["expires"]:raise ItemError("DECRYPTION_DENIED")
                except (PspError,ItemError) as exc:reject(p,exc.code)
        if operation=="process" and any(r["ok"] is not True for r in results):
            for r in results:
                if r["ok"] is True:reject({"output":r},"BATCH_REJECTED")
        successful=sum(r["ok"] is True for r in results)
        output={"profile":SECURITY_TOOLS_PROFILE,"operation_id":value["operation_id"],"policy_version":version,"success":len(results)>0 and successful==len(results),"results":results,"summary":{"total":len(results),"successful":successful,"failed":len(results)-successful},**({"untrustedText":bool(untrusted_text)} if operation!="decrypt" else {})}
        if len(canonical_json(output).encode("utf-8"))>MAX_REQUEST_BYTES:raise ServiceError("RESPONSE_TOO_LARGE",500)
        return output

File: implementations/typescript/packages/api-server/src/security-tools.ts
// SPDX-License-Identifier: Apache-2.0
import {createHash,createDecipheriv} from "node:crypto";
import {PspError,byteLength,canonicalJson,parseMarkup,record,sectionToEnvelope,validateJson,type Section,type Envelope} from "@psp-cdl/core";
import {verifyEnvelope} from "@psp-cdl/core/crypto";
import {evaluateBatch,type PolicyInput} from "@psp-cdl/cdl";
import {SecurityService,ServiceError,identifier,requestObject,scopeFor,MAX_REQUEST_BYTES,type ServiceHost,type Operation,type OperationSnapshot,type Principal} from "./service.js";
import type {WorkflowService} from "./workflow.js";

export const SECURITY_TOOLS_PROFILE="PSP-SECURITY-TOOLS-0.1";
export const MAX_PLAINTEXT_BYTES=65_536;
export interface DecryptionContext {
  operationId:string;policyVersion:string;envelopeDigest:string;keyId:string;sectionType:string;mode:string;
}
export interface DecryptionGrant extends DecryptionContext {
  tenantId:string;subjectId:string;algorithm:"aes-256-gcm";material:Uint8Array;status:"active"|"revoked";
  expires:number;requestingZone:number;sectionTypes:string[];modes:string[];applicationOnly:boolean;
}
export interface PlaintextContext {
  operationId:string;policyVersion:string;operation:"decrypt"|"process";
  outputs:Record<string,unknown>[];
}
export interface SecurityToolsHost extends ServiceHost {
  resolveDecryption(principal:Principal,context:DecryptionContext):DecryptionGrant|null|Promise<DecryptionGrant|null>;
  plaintextPolicy(principal:Principal,context:PlaintextContext):{allow:boolean;complete:boolean;resources:PolicyInput[]}|Promise<{allow:boolean;complete:boolean;resources:PolicyInput[]}>;
}
class ItemError extends Error {constructor(readonly code:string){super(code);}}
function fail(code:string):never {throw new ItemError(code);}
const digest=(e:Envelope)=>createHash("sha256").update(canonicalJson(e)).digest("hex");
function base64(value:unknown,length?:number):Buffer {
  if(typeof value!=="string"||value.length>4*Math.ceil(MAX_PLAINTEXT_BYTES/3)||value.length%4!==0||!/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(value))fail("INVALID_ENCRYPTION");
  const bytes=Buffer.from(value,"base64");
  if(bytes.toString("base64")!==value||length!==undefined&&bytes.length!==length)fail("INVALID_ENCRYPTION");
  return bytes;
}
function scan(source:unknown):{sections:Section[];untrustedText:boolean} {
  if(typeof source!=="string")throw new ServiceError("INVALID_REQUEST",400);
  const doc=parseMarkup(source),sections:Section[]=[];
  const visit=(section:Section)=>{sections.push(section);if(sections.length>32)throw new ServiceError("SECTION_LIMIT",413);for(const child of section.children)if(child.kind==="section")visit(child);};
  for(const child of doc.children)if(child.kind==="section")visit(child);
  return {sections,untrustedText:doc.children.some(n=>n.kind==="text"&&/[^\t\r\n ]/.test(n.value))};
}
function envelope(section:Section):Envelope {
  if(section.children.some(n=>n.kind==="section"))fail("UNSUPPORTED_NESTING");
  if(!Object.hasOwn(section.attributes,"signature"))fail(section.attributes.encrypted==="true"?"UNSUPPORTED_ENCRYPTION_ORDER":"UNSIGNED_SECTION");
  if(section.attributes.type==="user")fail("UNSUPPORTED_SECTION_TYPE");
  return sectionToEnvelope(section);
}
const encryptionFields=["encrypted","encryption-algorithm","encryption-key-id","nonce","tag"];
function encrypted(e:Envelope):boolean {
  return encryptionFields.some(k=>Object.hasOwn(e.signature.attributes??{},k))||record(e.data)&&(Object.hasOwn(e.data,"encryption")||Object.hasOwn(e.data,"encryptedData"));
}
function parameters(e:Envelope) {
  const attrs=e.signature.attributes??{},mode=attrs.decrypt??"upfront";
  if(!["system","context"].includes(e.signature.sectionType))fail("UNSUPPORTED_SECTION_TYPE");
  if(!["upfront","node","on-request"].includes(mode))fail("UNSUPPORTED_DECRYPT_MODE");
  let algorithm:unknown,keyId:unknown,nonce:unknown,tag:unknown,ciphertext:unknown;
  if(e.signature.contentType==="text") {
    if(attrs.encrypted!=="true"||encryptionFields.some(k=>!Object.hasOwn(attrs,k)))fail("INVALID_ENCRYPTION");
    algorithm=attrs["encryption-algorithm"];keyId=attrs["encryption-key-id"];nonce=attrs.nonce;tag=attrs.tag;ciphertext=e.data;
  } else {
    if(encryptionFields.some(k=>Object.hasOwn(attrs,k))||!record(e.data)||Object.keys(e.data).sort().join(",")!=="encryptedData,encryption"||!record(e.data.encryption)||Object.keys(e.data.encryption).sort().join(",")!=="algorithm,keyId,nonce,tag")fail("INVALID_ENCRYPTION");
    ({algorithm,keyId,nonce,tag}=e.data.encryption);ciphertext=e.data.encryptedData;
  }
  if(algorithm!=="aes-256-gcm")fail("UNSUPPORTED_ENCRYPTION_ALGORITHM");
  if(!identifier(keyId))fail("INVALID_ENCRYPTION");
  return {keyId,mode,nonce:base64(nonce,12),tag:base64(tag,16),ciphertext:base64(ciphertext)};
}
function allowed(resources:unknown):boolean {return Array.isArray(resources)&&resources.length>0&&evaluateBatch(resources).decision==="allow";}
type Prepared={output:Record<string,unknown>;envelope:Envelope;context?:DecryptionContext;grant?:DecryptionGrant};

/** Explicit host integration; plaintext never bypasses the buffered release decision. */
export class SecurityToolsService extends SecurityService {
  override readonly operations:readonly Operation[];
  constructor(private readonly toolsHost:SecurityToolsHost,private readonly workflow?:WorkflowService) {
    super(toolsHost);
    if([toolsHost.resolveDecryption,toolsHost.plaintextPolicy,toolsHost.authenticate,toolsHost.resolve,toolsHost.now].some(f=>typeof f!=="function"))throw new ServiceError("INVALID_CONFIGURATION",500);
    this.operations=["verify","evaluate","scan","decrypt","process",...(workflow?.operations.filter(o=>o!=="verify"&&o!=="evaluate")??[])];
  }
  override async invoke(operation:Operation,value:unknown,token:unknown,expectedIdentity?:Principal):Promise<Record<string,unknown>> {
    if(operation!=="scan"&&operation!=="decrypt"&&operation!=="process") {
      if(operation==="verify"||operation==="evaluate")return super.invoke(operation,value,token,expectedIdentity);
      if(this.workflow){
        const p=await this.authenticate(token);
        if(expectedIdentity&&(expectedIdentity.tenantId!==p.tenantId||expectedIdentity.subjectId!==p.subjectId)||!p.scopes.includes(scopeFor(operation)))throw new ServiceError("FORBIDDEN",403);
        return this.workflow.invoke(operation,value,token,p);
      }
      throw new ServiceError("UNSUPPORTED_OPERATION",404);
    }
    try {
      const principal=await this.authenticate(token),copyPrincipal=()=>validateJson(principal) as unknown as Principal;
      if(expectedIdentity&&(expectedIdentity.tenantId!==principal.tenantId||expectedIdentity.subjectId!==principal.subjectId))throw new ServiceError("FORBIDDEN",403);
      if(!principal.scopes.includes(scopeFor(operation)))throw new ServiceError("FORBIDDEN",403);
      const input=requestObject(value,operation==="decrypt"?["operation_id","sections"]:["operation_id","raw_text"]);
      const fresh=async(version?:string)=>{
        const live=await this.authenticate(token);
        if(live.tenantId!==principal.tenantId||live.subjectId!==principal.subjectId||!live.scopes.includes(scopeFor(operation)))throw new ServiceError("FORBIDDEN",403);
        let s:OperationSnapshot|null;try{s=structuredClone(await this.toolsHost.resolve(copyPrincipal(),input.operation_id as string));}catch{throw new ServiceError("INTERNAL_ERROR",500);}
        if(!s||s.tenantId!==principal.tenantId||s.subjectId!==principal.subjectId||s.operationId!==input.operation_id)throw new ServiceError("NOT_FOUND",404);
        const now=this.toolsHost.now();
        if(!Number.isFinite(now)||now<0||now>Number.MAX_SAFE_INTEGER||!Number.isSafeInteger(s.expires)||!identifier(s.policyVersion))throw new ServiceError("INTERNAL_ERROR",500);
        if(now>=s.expires||version!==undefined&&version!==s.policyVersion)throw new ServiceError("STALE_OPERATION",409);
        const required={"tenant-id":principal.tenantId,"operation-id":s.operationId,"policy-version":s.policyVersion},p=s.verification;
        if(!p||!record(p.context)||Object.entries(required).some(([k,v])=>p.context[k]!==v)||!Array.isArray(p.keys)||p.keys.some(k=>k.allowUnscoped!==false))throw new ServiceError("INTERNAL_ERROR",500);
        if(!allowed(s.resources))throw new ServiceError("POLICY_DENIED",403);
        return {...s,verification:{...p,now}};
      };
      const snapshot=await fresh(),version=snapshot.policyVersion;
      const grant=async(context:DecryptionContext):Promise<DecryptionGrant>=>{
        let g:DecryptionGrant|null;try{g=structuredClone(await this.toolsHost.resolveDecryption(copyPrincipal(),validateJson(context) as unknown as DecryptionContext));}catch{throw new ServiceError("INTERNAL_ERROR",500);}
        const target=context.sectionType==="system"?0:1;
        if(!g||g.tenantId!==principal.tenantId||g.subjectId!==principal.subjectId||Object.entries(context).some(([k,v])=>g![k as keyof DecryptionGrant]!==v)||g.algorithm!=="aes-256-gcm"||g.status!=="active"||!(g.material instanceof Uint8Array)||g.material.length!==32||!Number.isSafeInteger(g.expires)||this.toolsHost.now()>=g.expires||g.applicationOnly!==true||!Number.isInteger(g.requestingZone)||g.requestingZone<0||g.requestingZone>2||g.requestingZone>target||!Array.isArray(g.sectionTypes)||!g.sectionTypes.includes(context.sectionType)||!Array.isArray(g.modes)||!g.modes.includes(context.mode))fail("DECRYPTION_DENIED");
        return g;
      };
      let items:{id:string;section?:Section;content?:string}[],untrustedText=false;
      if(operation==="decrypt") {
        if(!Array.isArray(input.sections)||input.sections.length<1||input.sections.length>32)throw new ServiceError("INVALID_REQUEST",400);
        const ids=new Set<string>();items=input.sections.map(s=>{if(!record(s)||Object.keys(s).sort().join(",")!=="content,id"||!identifier(s.id)||typeof s.content!=="string"||ids.has(s.id))throw new ServiceError("INVALID_REQUEST",400);ids.add(s.id);return {id:s.id,content:s.content};});
      }else{
        const scanned=scan(input.raw_text);untrustedText=scanned.untrustedText;
        if(operation==="process"&&(untrustedText||!scanned.sections.length))throw new ServiceError("UNSUPPORTED_UNTAGGED_INPUT",400);
        items=scanned.sections.map((section,i)=>({id:"section-"+i,section}));
      }
      const results:Record<string,unknown>[]=[],prepared:Prepared[]=[];
      for(const item of items) {
        try {
          let section=item.section;
          if(!section){const doc=parseMarkup(item.content!);if(doc.children.length!==1||doc.children[0]!.kind!=="section")fail("INVALID_SECTION");section=doc.children[0] as Section;}
          const e=verifyEnvelope(envelope(section),snapshot.verification),isEncrypted=encrypted(e);
          if(operation==="scan"){const output={id:item.id,ok:true,encrypted:isEncrypted,sectionType:e.signature.sectionType};results.push(output);prepared.push({output,envelope:e});continue;}
          const provenance={profile:SECURITY_TOOLS_PROFILE,envelopeDigest:digest(e),signatureAlgorithm:e.signature.algorithm,signingKeyId:e.signature.kid??e.signature.secretId,trustLevel:e.signature.trustLevel??2,transformation:isEncrypted?"decrypted":"verified"};
          const candidate:Prepared={output:{id:item.id,ok:true,encrypted:isEncrypted,sectionType:e.signature.sectionType,provenance},envelope:e};
          let content:string;
          if(isEncrypted){
            const p=parameters(e),context={operationId:snapshot.operationId,policyVersion:version,envelopeDigest:provenance.envelopeDigest,keyId:p.keyId,sectionType:e.signature.sectionType,mode:p.mode};
            const g=await grant(context);candidate.context=context;candidate.grant=g;
            const beforeDecrypt=await fresh(version);verifyEnvelope(e,beforeDecrypt.verification);
            if(beforeDecrypt.verification.now>=g.expires)fail("DECRYPTION_DENIED");
            let bytes:Buffer;
            try{const decipher=createDecipheriv("aes-256-gcm",g.material,p.nonce,{authTagLength:16});decipher.setAuthTag(p.tag);bytes=Buffer.concat([decipher.update(p.ciphertext),decipher.final()]);}catch{fail("DECRYPTION_FAILED");}
            if(bytes.length>MAX_PLAINTEXT_BYTES)fail("PLAINTEXT_LIMIT");
            try{content=new TextDecoder("utf-8",{fatal:true,ignoreBOM:true}).decode(bytes);}catch{fail("INVALID_PLAINTEXT");}
          }else{
            if(operation==="decrypt")fail("NOT_ENCRYPTED");
            if(typeof e.data!=="string"||e.signature.contentType!=="text")fail("UNSUPPORTED_PLAINTEXT");content=e.data;
          }
          if(byteLength(content)>MAX_PLAINTEXT_BYTES)fail("PLAINTEXT_LIMIT");
          candidate.output.content=content;prepared.push(candidate);results.push(candidate.output);
        }catch(e){if(e instanceof PspError||e instanceof ItemError)results.push({id:item.id,ok:false,error:e.code});else throw e;}
      }
      if(operation!=="scan"&&prepared.length&&(operation!=="process"||prepared.length===results.length)) {
        await fresh(version);
        let release;try{release=validateJson(await this.toolsHost.plaintextPolicy(copyPrincipal(),{operationId:snapshot.operationId,policyVersion:version,operation,outputs:validateJson(prepared.map(p=>p.output)) as Record<string,unknown>[]}));}catch{throw new ServiceError("INTERNAL_ERROR",500);}
        const permit=record(release)&&release.allow===true&&release.complete===true&&allowed(release.resources);
        for(const p of prepared) {
          try{
            if(!permit)fail("PLAINTEXT_DENIED");
            if(p.context){const latest=await grant(p.context);if(!Buffer.from(latest.material).equals(Buffer.from(p.grant!.material)))fail("DECRYPTION_DENIED");p.grant=latest;}
          }catch(e){if(!(e instanceof ItemError))throw e;delete p.output.content;delete p.output.provenance;p.output.ok=false;p.output.error=e.code;}
        }
      }
      const final=await fresh(version);
      for(const p of prepared)if(p.output.ok===true){try{verifyEnvelope(p.envelope,final.verification);if(p.grant&&final.verification.now>=p.grant.expires)fail("DECRYPTION_DENIED");}catch(e){if(!(e instanceof PspError||e instanceof ItemError))throw e;delete p.output.content;delete p.output.provenance;p.output.ok=false;p.output.error=e.code;}}
      if(operation==="process"&&results.some(r=>r.ok!==true))for(const r of results)if(r.ok===true){delete r.content;delete r.provenance;r.ok=false;r.error="BATCH_REJECTED";}
      const successful=results.filter(r=>r.ok===true).length;
      const output={profile:SECURITY_TOOLS_PROFILE,operation_id:input.operation_id,policy_version:version,success:results.length>0&&successful===results.length,results,summary:{total:results.length,successful,failed:results.length-successful},...(operation!=="decrypt"?{untrustedText}:{})};
      if(byteLength(canonicalJson(output))>MAX_REQUEST_BYTES)throw new ServiceError("RESPONSE_TOO_LARGE",500);
      return output;
    }catch(e){if(e instanceof ServiceError)throw e;if(e instanceof PspError)throw new ServiceError(e.code,400);throw new ServiceError("INTERNAL_ERROR",500);}
  }
}
